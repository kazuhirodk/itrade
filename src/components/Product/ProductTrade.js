import React, { Component } from 'react'
import Swiper from 'react-native-deck-swiper'
import { Alert, Image, Button, StyleSheet, Text, View } from 'react-native'
import FirebaseService from '../../../services/FirebaseService';
import firebase from 'react-native-firebase';
import { Actions } from 'react-native-router-flux';

export default class ProductTrade extends Component {
  constructor (props) {
    super(props)
    this.state = {
      cards: [{id: 'teste', name: 'Clique no card para iniciar'}],
      swipedAllCards: false,
      swipeDirection: '',
      cardIndex: 0,
      allProducts: [],
      offeredProduct: '',
      interestedProduct: '',
      match: {
        myProduct: {
          contato: '',
          id: '',
          name: '',
          ownerName: '',
          sourceImage: ''
        },
        interestProduct: {
          contato: '',
          id: '',
          name: '',
          ownerName: '',
          sourceImage: ''
        }
      }
    }
  }

  key = ''
  user = ''

  componentDidMount(){
    firebase.auth().onAuthStateChanged(userLogged => {
      const allUsers = FirebaseService.getDataList('usuarios', function(){});

      allUsers.orderByChild("email").equalTo(userLogged.email).on("child_added", snapshot => {
        key = snapshot.key;
        user = snapshot.val();

        myProductInfo = this.state.match.myProduct;

        produto = user.produtos['-Lqayd0OzBYMZE1H3Ota'];

        console.log(this.state)
      })

      allUsers.orderByChild("email").on("child_added", snapshot => {
        let generalUser = snapshot.val();

        if(typeof generalUser.produtos !== "undefined" && userLogged.email !== generalUser.email){
          var allProductsArray = this.state['cards'];
          var produtos = generalUser.produtos;
          var keys = Object.keys(produtos);

          keys.forEach(function(key){
            var index = allProductsArray.findIndex( x => x.id==key);

            if (index === -1) {
              allProductsArray.push({id: key, name: produtos[key].nome, sourceImage: produtos[key].foto, ownerName: generalUser.nome_usuario, ownerContact: generalUser.telefone});
            } else console.log('object already exists')
          });

          if (this.state.cards == [{id: 'teste', name: 'Clique no card para iniciar'}]){ //gambiarra monstra pra arrumar o indice, depois eu arrumo
            this.setState({
              cards: [{id: 'teste', name: 'Clique no card para iniciar'}].concat(allProductsArray)
            })
          }
        }
      })
    })
  }

  renderCard = (card, index) => {
    return (
      <View style={styles.card}>
        <Image style={{width: 150, height: 150}} source={{uri: card.sourceImage}}/>
        <Text style={styles.text}>{card.name}</Text>
      </View>
    )
  };

  onSwiped = (type) => {
    console.log(`on swiped ${type}`)
  }

  onSwipedAllCards = () => {
    this.setState({
      swipedAllCards: true
    })
    Alert.alert('Os produtos na sua região acabaram :(', 'Tente trocar novamente mais tarde');
    Actions.home();
  };

  swipeLeft = (cardIndex) => {
    this.setState({
      offeredProduct: '-Lqayd0OzBYMZE1H3Ota',
      interestedProduct: this.state.cards[cardIndex].id,
      cardIndex: cardIndex
    })

    this.swiper.swipeLeft()
  };

  swipeRight = (cardIndex) => {
    this.setState({
      offeredProduct: '-Lqayd0OzBYMZE1H3Ota',
      interestedProduct: this.state.cards[cardIndex].id,
      cardIndex: cardIndex
    })

    FirebaseService.pushData('likes', {produto_interesse: this.state.cards[cardIndex].id, produto_oferta: this.state.offeredProduct})

    const likesOnMe = FirebaseService.getDataList('likes', function(){});

    likesOnMe.orderByChild('produto_interesse').equalTo(this.state.offeredProduct).on('child_added', snapshot => {
      let match = snapshot.val();

      if(match.produto_oferta == this.state.cards[cardIndex].id) {
        this.setState({
          match: {
            myProduct: {
              contato: user.telefone,
              id: '-Lqayd0OzBYMZE1H3Ota',
              name: produto.nome,
              ownerName: user.nome_usuario,
              sourceImage: produto.foto
            },
            interestProduct: {
              contato: this.state.cards[cardIndex].ownerContact,
              id: match.produto_oferta,
              name: this.state.cards[cardIndex].name,
              ownerName: this.state.cards[cardIndex].ownerName,
              sourceImage: this.state.cards[cardIndex].sourceImage
            }
          }
        })

        FirebaseService.pushData('usuarios/' + key + '/matches', this.state.match);

        Alert.alert('Você deu match!', 'Clique em OK para continuar',
          [
            {text: 'OK', onPress: () => console.log('DEU MATCH!')},
          ],
          {cancelable: false},
        )
      }
    })
  };

  render () {
    return (
      <View style={styles.container}>
        <Swiper
          ref={swiper => {
            this.swiper = swiper
          }}
          // onSwiped={() => this.onSwiped('general')}
          // onSwipedLeft={() => this.onSwiped('left')}
          onSwipedRight={this.swipeRight}
          // onSwipedTop={() => this.onSwiped('top')}
          // onSwipedBottom={() => this.onSwiped('bottom')}
          onTapCard={this.swipeLeft}
          cards={this.state.cards}
          cardIndex={this.state.cardIndex}
          cardVerticalMargin={80}
          renderCard={this.renderCard}
          onSwipedAll={this.onSwipedAllCards}
          stackSize={3}
          stackSeparation={15}
          overlayLabels={{
            bottom: {
              title: 'PULAR',
              style: {
                label: {
                  backgroundColor: 'black',
                  borderColor: 'black',
                  color: 'white',
                  borderWidth: 1
                },
                wrapper: {
                  flexDirection: 'column',
                  alignItems: 'center',
                  justifyContent: 'center'
                }
              }
            },
            left: {
              title: 'NOPE',
              style: {
                label: {
                  backgroundColor: 'black',
                  borderColor: 'black',
                  color: 'white',
                  borderWidth: 1
                },
                wrapper: {
                  flexDirection: 'column',
                  alignItems: 'flex-end',
                  justifyContent: 'flex-start',
                  marginTop: 30,
                  marginLeft: -30
                }
              }
            },
            right: {
              title: 'LIKE',
              style: {
                label: {
                  backgroundColor: 'black',
                  borderColor: 'black',
                  color: 'white',
                  borderWidth: 1
                },
                wrapper: {
                  flexDirection: 'column',
                  alignItems: 'flex-start',
                  justifyContent: 'flex-start',
                  marginTop: 30,
                  marginLeft: 30
                }
              }
            },
            top: {
              title: 'SUPER LIKE',
              style: {
                label: {
                  backgroundColor: 'black',
                  borderColor: 'black',
                  color: 'white',
                  borderWidth: 1
                },
                wrapper: {
                  flexDirection: 'column',
                  alignItems: 'center',
                  justifyContent: 'center'
                }
              }
            }
          }}
          animateOverlayLabelsOpacity
          animateCardOpacity
          swipeBackCard
        >
        </Swiper>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F5FCFF'
  },
  card: {
    flex: 1,
    borderRadius: 4,
    borderWidth: 2,
    borderColor: '#E8E8E8',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',
    height: '50%'
  },
  text: {
    textAlign: 'center',
    fontSize: 50,
    backgroundColor: 'transparent'
  },
  done: {
    textAlign: 'center',
    fontSize: 30,
    color: 'white',
    backgroundColor: 'transparent'
  }
})
