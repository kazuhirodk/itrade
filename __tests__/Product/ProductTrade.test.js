import ProductTrade from '../../src/components/Product/ProductTrade.js'
import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow } from 'enzyme';
import { Image, Text, View } from 'react-native'

import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

const wrapper = shallow(<ProductTrade />);

describe('#ProductTrade', () => {
  it('matches snapshot properly', () => {
    const jsonProductTrade = renderer.create(<ProductTrade />).toJSON();
    expect(jsonProductTrade).toMatchSnapshot();
  });

  describe('#renderCard', () => {
    it('render card properly', () => {
      const renderedResponse = <View style={{"alignItems": "center", "backgroundColor": "white", "borderColor": "#E8E8E8", "borderRadius": 4, "borderWidth": 2, "flex": 1, "height": "50%", "justifyContent": "center"}}><Image source={{"testUri": "../../../src/components/images/product1.png"}} style={{"height": 150, "width": 150}} /><Text style={{"backgroundColor": "transparent", "fontSize": 50, "textAlign": "center"}} /></View>
      expect(wrapper.instance().renderCard()).toEqual(renderedResponse);
    })
  });

  describe('onSwipeAllCards', () => {
    it('set state for swipedAllCards to true', () => {
      wrapper.instance().onSwipedAllCards();
      expect(wrapper.instance().state.swipedAllCards).toBeTruthy();
    })
  });
});
