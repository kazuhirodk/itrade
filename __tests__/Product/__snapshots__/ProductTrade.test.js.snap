// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ProductTrade matchs snapshot properly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "backgroundColor": "#4FD0E9",
          "marginBottom": 0,
          "marginTop": 0,
        },
        Object {},
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "zIndex": -3,
          },
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Desfazer ação
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "translateX": -750,
            },
            Object {
              "translateY": -1334,
            },
          ],
          "width": 710,
          "zIndex": 4,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#E8E8E8",
            "borderRadius": 4,
            "borderWidth": 2,
            "flex": 1,
            "height": "50%",
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/components/images/product1.png",
            }
          }
          style={
            Object {
              "height": 150,
              "width": 150,
            }
          }
        />
        <Text
          style={
            Object {
              "backgroundColor": "transparent",
              "fontSize": 50,
              "textAlign": "center",
            }
          }
        >
          produto 2
        </Text>
      </View>
    </View>
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "opacity": 1,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
            Object {
              "rotate": "0deg",
            },
          ],
          "width": 710,
          "zIndex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#E8E8E8",
            "borderRadius": 4,
            "borderWidth": 2,
            "flex": 1,
            "height": "50%",
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/components/images/product1.png",
            }
          }
          style={
            Object {
              "height": 150,
              "width": 150,
            }
          }
        />
        <Text
          style={
            Object {
              "backgroundColor": "transparent",
              "fontSize": 50,
              "textAlign": "center",
            }
          }
        >
          produto 1
        </Text>
      </View>
    </View>
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
          "height": 1174,
          "left": 20,
          "position": "absolute",
          "top": 80,
          "transform": Array [
            Object {
              "scale": 0.97,
            },
            Object {
              "translateY": 15,
            },
          ],
          "width": 710,
          "zIndex": -1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#E8E8E8",
            "borderRadius": 4,
            "borderWidth": 2,
            "flex": 1,
            "height": "50%",
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/components/images/product1.png",
            }
          }
          style={
            Object {
              "height": 150,
              "width": 150,
            }
          }
        />
        <Text
          style={
            Object {
              "backgroundColor": "transparent",
              "fontSize": 50,
              "textAlign": "center",
            }
          }
        >
          produto 2
        </Text>
      </View>
    </View>
  </View>
</View>
`;
